### PR TITLE
Fix issue where nagative hits are put in the cache as null and it NPEs

### PR DESCRIPTION
**Pull request by i386**
# Description

See [JENKINS-XXXXX](https://issues.jenkins-ci.org/browse/JENKINS-XXXXX).

# Submitter checklist
- [ ] Link to JIRA ticket in description, if appropriate.
- [ ] Change is code complete and matches issue description
- [ ] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

### CODE REVIEW

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/BranchImpl.java [hunk 1]
```diff
@@ -24,9 +24,7 @@
 
 import java.util.concurrent.ExecutionException;
 
-import static io.jenkins.blueocean.rest.model.KnownCapabilities.BLUE_BRANCH;
-import static io.jenkins.blueocean.rest.model.KnownCapabilities.JENKINS_WORKFLOW_JOB;
-import static io.jenkins.blueocean.rest.model.KnownCapabilities.PULL_REQUEST;
+import static io.jenkins.blueocean.rest.model.KnownCapabilities.*;
```
Comment (michaelneale) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/BranchImpl.java]: I thought the usual standard is for explicit imports...

**Comment by michaelneale** [comment]
failure could be legit: 

```
testPullRequestCacheLoaderWithoutScmHead – io.jenkins.blueocean.rest.impl.pipeline.CachesTest
a few seconds
Stacktrace
java.lang.NullPointerException
	at io.jenkins.blueocean.rest.impl.pipeline.CachesTest.testPullRequestCacheLoaderWithoutScmHead(CachesTest.java:131)
```

**Comment by michaelneale** [comment]
@i386 still relevant? might need a little more massaging

**Comment by i386** [comment]
@michaelneale I just fixed the test and updated the branch

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/Caches.java [hunk 2]
```diff
@@ -30,17 +31,32 @@
...
 
 
-    static final LoadingCache<String, Branch> BRANCH_METADATA = CacheBuilder.newBuilder()
+    static final LoadingCache<String, CacheValue<Branch>> BRANCH_METADATA = CacheBuilder.newBuilder()
             .maximumSize(BRANCH_METADATA_CACHE_MAX_SIZE)
             .expireAfterAccess(1, TimeUnit.DAYS)
             .build(new BranchCacheLoader(Jenkins.getInstance().getItemGroup()));
 
+    /** Generic cache value to avoid putting negative hits as nulls in the cache */
+    static class CacheValue<T> {
```
Comment (vivek) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/Caches.java]: You could use guava `Optional` instead?

Comment (i386) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/Caches.java]: Done :)

**Comment by i386** [comment]
gah was terminated. this PR will never end :(

**Comment by michaelneale** [comment]
the last few test failures look legit - will see how this one goes.

**Pull request merged**